src/Workspaces/MSBuildTest/OpenProjectsTests.cs

```diff
@@ -0,0 +1,225 @@
+
+            static string GenerateProjectFromTemplate(string projectPath, string templateName, string languageName, ITestOutputHelper outputHelper)
+            {
+                var projectFilePath = GetProjectFilePath(projectPath, languageName);
+
+                CreateNewProject(templateName, projectPath, languageName, outputHelper);
+
+                return projectFilePath;
+            }
+
+            static string GetProjectFilePath(string projectPath, string languageName)
+            {
+                var projectName = Path.GetFileName(projectPath);
+                var projectExtension = languageName switch
+                {
+                    LanguageNames.CSharp => "csproj",
+                    LanguageNames.VisualBasic => "vbproj",
+                    _ => throw new ArgumentOutOfRangeException(nameof(languageName), actualValue: languageName, message: "Only C# and VB.Net project are supported.")
+                };
+                return Path.Combine(projectPath, $"{projectName}.{projectExtension}");
+            }
+
+            static int CreateNewProject(string templateName, string outputPath, string languageName, ITestOutputHelper output)
+            {
+                var language = languageName switch
+                {
+                    LanguageNames.CSharp => "C#",
+                    LanguageNames.VisualBasic => "VB",
+                    _ => throw new ArgumentOutOfRangeException(nameof(languageName), actualValue: languageName, message: "Only C#, F# and VB.NET project are supported.")
+                };
+
+                var newResult = RunDotNet($"new \"{templateName}\" -o \"{outputPath}\" --language \"{language}\"");
+                Assert.Equal(0, newResult.ExitCode);
+
+                output.WriteLine(string.Join(Environment.NewLine, newResult.Output));
+
+                // Most templates invoke restore as a post-creation action. However, some, like the
+                // Maui templates, do not run restore since they require additional workloads to be
+                // installed.
+                if (newResult.Output.Contains("Restoring"))
+                {
+                    return newResult.ExitCode;
+                }
+
+                // Attempt a restore and see if we are instructed to install additional workloads.
+                var restoreResult = RunDotNet($"restore", workingDirectory: outputPath);
+
+                output.WriteLine(string.Join(Environment.NewLine, restoreResult.Output));
+
+                if (restoreResult.ExitCode == 0)
+                    return restoreResult.ExitCode;
+
+                if (restoreResult.Output.Contains("command: dotnet workload restore"))
+                    throw new InvalidOperationException($"Since the '{templateName}' template requires additional dotnet workloads to be installed, it should likely be excluded during template discovery.");
+
+                throw new InvalidOperationException($"The dotnet restore operation failed.");
```

Comment:
Should this include the process text in the failure? Sometimes the test output is a bit hard to fish out from CI.